airbyte-integrations/connectors/destination-kafka/src/main/java/io/airbyte/integrations/destination/kafka/KafkaRecordConsumer.java

```diff
@@ -0,0 +1,156 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.destination.kafka;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableMap;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.base.AirbyteStreamNameNamespacePair;
+import io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer;
+import io.airbyte.integrations.base.JavaBaseConstants;
+import io.airbyte.integrations.destination.NamingConventionTransformer;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.Future;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.KafkaException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KafkaRecordConsumer extends FailureTrackingAirbyteMessageConsumer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(KafkaRecordConsumer.class);
+
+  private final String topicPattern;
+  private final Map<AirbyteStreamNameNamespacePair, String> topicMap;
+  private final KafkaProducer<String, JsonNode> producer;
+  private final boolean sync;
+  private final boolean transactionalProducer;
+  private final ConfiguredAirbyteCatalog catalog;
+  private final Callback callback;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
+  private final NamingConventionTransformer nameTransformer;
+
+  private AirbyteMessage lastStateMessage = null;
+
+  public KafkaRecordConsumer(KafkaDestinationConfig kafkaDestinationConfig,
+                             ConfiguredAirbyteCatalog catalog,
+                             Consumer<AirbyteMessage> outputRecordCollector,
+                             NamingConventionTransformer nameTransformer) {
+    this.topicPattern = kafkaDestinationConfig.getTopicPattern();
+    this.topicMap = new HashMap<>();
+    this.producer = kafkaDestinationConfig.getProducer();
+    this.sync = kafkaDestinationConfig.isSync();
+    this.transactionalProducer = kafkaDestinationConfig.isTransactionalProducer();
+    this.catalog = catalog;
+    this.callback = (metadata, exception) -> {
+      if (exception != null) {
+        // TODO improve error management
+        LOGGER.error("Error sending message to topic", exception);
```

Comment:
this constitutes data loss -- shouldn't it fail the sync?

Reply:
Yeah, I just changed the callback so it'll throw a ``RuntimeException`` and will cause the destination to fail